File: services/auth/oidc.py
"""
Google OpenID Connect (OIDC) Module — handles Google OAuth2 authorization URL
generation, ID token decoding/validation, and user claims parsing.
"""

import base64
import json
import secrets
import time
from urllib.parse import urlencode

import httpx

from .config import (
    GOOGLE_AUTH_ENDPOINT,
    GOOGLE_CLIENT_ID,
    GOOGLE_CLIENT_SECRET,
    GOOGLE_REDIRECT_URI,
    GOOGLE_TOKEN_ENDPOINT,
)


# region Authorization URL & State Helper
def generate_state_token() -> str:
    """Generate a cryptographically secure random state token for CSRF mitigation."""
    return secrets.token_urlsafe(32)


def build_google_auth_url(
    client_id: str = GOOGLE_CLIENT_ID,
    redirect_uri: str = GOOGLE_REDIRECT_URI,
    state: str | None = None,
) -> str:
    """Generate the Google OAuth 2.0 / OIDC authorization redirect URL."""
    if state is None:
        state = generate_state_token()

    params = {
        "client_id": client_id,
        "redirect_uri": redirect_uri,
        "response_type": "code",
        "scope": "openid email profile",
        "access_type": "offline",
        "state": state,
        "prompt": "consent",
    }
    return f"{GOOGLE_AUTH_ENDPOINT}?{urlencode(params)}"


# endregion


# region OAuth 2.0 Authorization Code Exchange
async def exchange_code_for_id_token(
    code: str,
    redirect_uri: str = GOOGLE_REDIRECT_URI,
    client_id: str = GOOGLE_CLIENT_ID,
    client_secret: str = GOOGLE_CLIENT_SECRET,
) -> str:
    """Exchange an OAuth 2.0 authorization code for a Google ID Token.

    Supports mock codes for offline unit and E2E testing environments.
    """
    if not code:
        raise ValueError("Authorization code must not be empty")

    # Handle mock authorization code for offline test environments
    if code.startswith("mock_code_") or client_id.startswith("mock-"):
        return build_mock_google_id_token(
            email=f"user_{code[-6:]}@gmail.com" if len(code) >= 6 else "mock_google@gmail.com",
            sub=f"google_sub_{code}",
            name="Mock Google User",
        )

    payload = {
        "code": code,
        "client_id": client_id,
        "client_secret": client_secret,
        "redirect_uri": redirect_uri,
        "grant_type": "authorization_code",
    }

    async with httpx.AsyncClient(timeout=10.0) as client:
        response = await client.post(GOOGLE_TOKEN_ENDPOINT, data=payload)
        if response.status_code != 200:
            raise ValueError(f"Failed to exchange code with Google Token Endpoint: {response.text}")

        data = response.json()
        id_token = data.get("id_token")
        if not id_token:
            raise ValueError("Google token endpoint response did not include 'id_token'")
        return id_token


# endregion


# region Google ID Token Parser & Validator
def parse_google_id_token(id_token: str) -> dict:
    """Decode and validate a Google OIDC ID token (JWT format).

    Extracts user email, google_sub, name, and profile picture.
    Supports mock tokens for offline testing.
    """
    if not id_token or not isinstance(id_token, str):
        raise ValueError("Invalid Google ID Token format")

    parts = id_token.strip().split(".")

    # Handle mock raw JSON payload for unit testing flexibility
    if len(parts) == 1 and id_token.startswith("{"):
        try:
            payload = json.loads(id_token)
        except Exception as err:
            raise ValueError(f"Invalid JSON in ID Token: {err}")
    elif len(parts) == 3:
        # Standard JWT (header.payload.signature)
        payload_b64 = parts[1]
        # Pad Base64 if needed
        rem = len(payload_b64) % 4
        if rem > 0:
            payload_b64 += "=" * (4 - rem)
        try:
            decoded_bytes = base64.urlsafe_b64decode(payload_b64)
            payload = json.loads(decoded_bytes.decode("utf-8"))
        except Exception as err:
            raise ValueError(f"Failed to decode Google ID Token payload: {err}")
    else:
        raise ValueError("Malformed Google ID Token format (expected JWT header.payload.signature)")

    email = payload.get("email")
    sub = payload.get("sub")

    if not email or not sub:
        raise ValueError("Google ID Token payload missing required 'email' or 'sub' claims")

    return {
        "email": email,
        "google_sub": str(sub),
        "name": payload.get("name", email.split("@")[0]),
        "picture": payload.get("picture", ""),
        "payload": payload,
    }


# endregion


# region Mock Fixtures for Testing
def build_mock_google_id_token(
    email: str = "user@gmail.com",
    sub: str = "google_user_sub_123456789",
    name: str = "Google User",
) -> str:
    """Construct a Base64 URL-encoded mock Google ID Token for offline unit and E2E testing."""
    header = {"alg": "RS256", "typ": "JWT", "kid": "mock_google_key"}
    payload = {
        "iss": "https://accounts.google.com",
        "azp": GOOGLE_CLIENT_ID,
        "aud": GOOGLE_CLIENT_ID,
        "sub": sub,
        "email": email,
        "email_verified": True,
        "name": name,
        "picture": "https://lh3.googleusercontent.com/a/mock_photo",
        "iat": int(time.time()),
        "exp": int(time.time()) + 3600,
    }

    header_b64 = (
        base64.urlsafe_b64encode(json.dumps(header).encode("utf-8")).decode("utf-8").rstrip("=")
    )
    payload_b64 = (
        base64.urlsafe_b64encode(json.dumps(payload).encode("utf-8")).decode("utf-8").rstrip("=")
    )
    signature_b64 = (
        base64.urlsafe_b64encode(b"mock_google_rsa_signature").decode("utf-8").rstrip("=")
    )

    return f"{header_b64}.{payload_b64}.{signature_b64}"


# endregion

File: internal/auth/oidc.go
package auth

import (
	"context"
	"crypto/rand"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
	"time"
)

const googleAuthEndpoint = "https://accounts.google.com/o/oauth2/v2/auth"

type GoogleUser struct {
	Email     string
	GoogleSub string
	Name      string
	Picture   string
}

// GenerateStateToken returns a cryptographically random URL-safe OAuth state value.
func GenerateStateToken() (string, error) {
	data := make([]byte, 32)
	if _, err := rand.Read(data); err != nil {
		return "", err
	}
	return base64.RawURLEncoding.EncodeToString(data), nil
}

// BuildGoogleAuthURL constructs a Google OAuth authorization URL.
//
// clientID and redirectURI are copied into the query without validation. When
// state is empty, the function generates a cryptographically random state value;
// callers must retain that value from the returned URL and verify it during the
// callback. An error is returned only when state generation fails.
func BuildGoogleAuthURL(clientID, redirectURI, state string) (string, error) {
	if state == "" {
		var err error
		state, err = GenerateStateToken()
		if err != nil {
			return "", err
		}
	}
	query := url.Values{
		"client_id":     {clientID},
		"redirect_uri":  {redirectURI},
		"response_type": {"code"},
		"scope":         {"openid email profile"},
		"access_type":   {"offline"},
		"state":         {state},
		"prompt":        {"consent"},
	}
	return googleAuthEndpoint + "?" + query.Encode(), nil
}

// ExchangeGoogleCode exchanges a one-time Google authorization code for an ID token.
//
// code must be non-empty. redirectURI must match the URI used for authorization;
// when it is empty, cfg.GoogleCallbackURL is used. The function submits the code
// and client credentials to cfg.GoogleTokenURL with client, limits the response
// body to 1 MiB, and returns the id_token field from a successful response.
// Transport failures, non-200 responses, malformed JSON, and a missing ID token
// are returned as errors.
//
// When mock OIDC is explicitly enabled, the function performs no network request
// and returns a locally generated test token derived from code.
func ExchangeGoogleCode(ctx context.Context, client *http.Client, cfg Config, code, redirectURI string) (string, error) {
	if code == "" {
		return "", errors.New("authorization code must not be empty")
	}
	if isMockGoogleConfig(cfg) {
		email := "mock_google@gmail.com"
		if len(code) >= 6 {
			email = "user_" + code[len(code)-6:] + "@gmail.com"
		}
		return BuildMockGoogleIDToken(cfg.GoogleClientID, email, "google_sub_"+code, "Mock Google User")
	}
	if redirectURI == "" {
		redirectURI = cfg.GoogleCallbackURL
	}
	form := url.Values{
		"code":          {code},
		"client_id":     {cfg.GoogleClientID},
		"client_secret": {cfg.GoogleClientSecret},
		"redirect_uri":  {redirectURI},
		"grant_type":    {"authorization_code"},
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, cfg.GoogleTokenURL, strings.NewReader(form.Encode()))
	if err != nil {
		return "", err
	}
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	response, err := client.Do(req)
	if err != nil {
		return "", err
	}
	defer response.Body.Close()
	body, err := io.ReadAll(io.LimitReader(response.Body, 1<<20))
	if err != nil {
		return "", err
	}
	if response.StatusCode != http.StatusOK {
		return "", fmt.Errorf("failed to exchange code with Google Token Endpoint: %s", strings.TrimSpace(string(body)))
	}
	var payload struct {
		IDToken string `json:"id_token"`
	}
	if err := json.Unmarshal(body, &payload); err != nil {
		return "", err
	}
	if payload.IDToken == "" {
		return "", errors.New("Google token endpoint response did not include 'id_token'")
	}
	return payload.IDToken, nil
}

// ParseMockGoogleIDToken extracts a GoogleUser from a locally generated mock token.
//
// It requires a three-part JWT-shaped value, decodes the payload, and validates
// the email, subject, and email_verified claims. Missing names are derived from
// the email address. The placeholder signature, issuer, audience, and expiry are
// intentionally not verified, so this function must only be used behind the
// explicit mock-OIDC configuration guard.
func ParseMockGoogleIDToken(idToken string) (GoogleUser, error) {
	if strings.TrimSpace(idToken) == "" {
		return GoogleUser{}, errors.New("invalid Google ID Token format")
	}
	parts := strings.Split(strings.TrimSpace(idToken), ".")
	if len(parts) != 3 {
		return GoogleUser{}, errors.New("malformed Google ID Token format (expected JWT header.payload.signature)")
	}
	raw, err := base64.RawURLEncoding.DecodeString(parts[1])
	if err != nil {
		return GoogleUser{}, fmt.Errorf("failed to decode Google ID Token payload: %w", err)
	}
	var payload map[string]any
	if err := json.Unmarshal(raw, &payload); err != nil {
		return GoogleUser{}, fmt.Errorf("invalid JSON in ID Token: %w", err)
	}
	email, emailOK := payload["email"].(string)
	subValue, subOK := payload["sub"]
	sub := ""
	if subOK {
		sub = fmt.Sprint(subValue)
	}
	if !emailOK || email == "" || !subOK || sub == "" {
		return GoogleUser{}, errors.New("Google ID Token payload missing required 'email' or 'sub' claims")
	}
	name, _ := payload["name"].(string)
	if name == "" {
		name = strings.Split(email, "@")[0]
	}
	picture, _ := payload["picture"].(string)
	verified, _ := payload["email_verified"].(bool)
	if !verified {
		return GoogleUser{}, errors.New("Google ID Token email is not verified")
	}
	return GoogleUser{Email: email, GoogleSub: sub, Name: name, Picture: picture}, nil
}

// isMockGoogleConfig reports whether the explicit mock OIDC safeguards are both enabled.
func isMockGoogleConfig(cfg Config) bool {
	return cfg.AllowMockOIDC && strings.HasPrefix(cfg.GoogleClientID, "mock-")
}

// BuildMockGoogleIDToken creates a test-only Google-shaped ID token.
//
// The token contains the supplied audience and identity claims, a one-hour
// lifetime, and a placeholder signature that is not cryptographically valid.
// It must never be accepted by the production GoogleVerifier.
func BuildMockGoogleIDToken(clientID, email, sub, name string) (string, error) {
	header, err := json.Marshal(map[string]any{"alg": "RS256", "typ": "JWT", "kid": "mock_google_key"})
	if err != nil {
		return "", err
	}
	now := time.Now().Unix()
	payload, err := json.Marshal(map[string]any{
		"iss": "https://accounts.google.com", "azp": clientID, "aud": clientID,
		"sub": sub, "email": email, "email_verified": true, "name": name,
		"picture": "https://lh3.googleusercontent.com/a/mock_photo", "iat": now, "exp": now + 3600,
	})
	if err != nil {
		return "", err
	}
	signature := base64.RawURLEncoding.EncodeToString([]byte("mock_google_rsa_signature"))
	return base64.RawURLEncoding.EncodeToString(header) + "." + base64.RawURLEncoding.EncodeToString(payload) + "." + signature, nil
}
